tests/test_parallel_graceful.py

```diff
@@ -0,0 +1,223 @@
+
+
+@config.when(test_front="good")
+def input_maker__good(n: int) -> int:
+    return n
+
+
+@config.when(test_front="bad")
+def input_maker__bad(n: int) -> int:
+    raise Exception("Went wrong")
+
+
+@config.when(test_state="middle")
+def distro__middle(input_maker: int) -> Parallelizable[int]:
+    for x in range(input_maker):
+        if x > 4:
+            raise Exception("bad")
+        yield x * 3
+
+
+@config.when(test_state="early")
+def distro__early(input_maker: int) -> Parallelizable[int]:
+    raise Exception("bad")
+    for x in range(input_maker):
+        yield x * 3
+
+
+@config.when(test_state="pass")
+def distro__pass(input_maker: int) -> Parallelizable[int]:
+    for x in range(input_maker):
+        yield x * 3
+
+
+def some_math(distro: int) -> float:
+    if distro > 15:
+        raise Exception("No no no")
+    return distro * 2.0
+
+
+def other_math(distro: int) -> float:
+    if distro < 10:
+        raise Exception("Not allowed")
+    return distro + 1
+
+
+@tag(allow_injection="allow")
+def gather_math(some_math: float, other_math: float) -> List[float]:
+    return [some_math, other_math]
+
+
+def distro_end(gather_math: Collect[List[float]]) -> List[float]:
+    ans = [x for x in gather_math]
+    return ans
+
+
+def distro_gather(some_math: Collect[float]) -> List[float]:
+    ans = [x for x in some_math]
+    return ans
+
+
+# TODO: How does this break extract_fields, e.g.?
+
+temp_module_simple = ad_hoc_utils.create_temporary_module(
+    input_maker__good,
+    input_maker__bad,
+    distro__early,
+    distro__middle,
+    distro__pass,
+    some_math,
+    distro_gather,
+    module_name="parallel_example",
+)
+
+temp_module = ad_hoc_utils.create_temporary_module(
+    input_maker__good,
+    input_maker__bad,
+    distro__early,
+    distro__middle,
+    distro__pass,
+    some_math,
+    other_math,
+    gather_math,
+    distro_end,
+    module_name="parallel_example_complex",
+)
+
+
+def _make_driver(
+    module, test_state: str, fail_parallel=False, fail_first=False, injection_tags=None
+):
+    local_executor = SynchronousLocalTaskExecutor()
+    dr = (
+        driver.Builder()
+        .enable_dynamic_execution(allow_experimental_mode=True)
+        .with_modules(module)
+        .with_remote_executor(local_executor)
+        .with_adapters(
+            GracefulErrorAdapter(
+                error_to_catch=Exception,
+                sentinel_value=None,
+                fail_all_parallel=fail_parallel,
+                sentinel_injection_tags=injection_tags,
+            )
+        )
+        .with_config(
+            {
+                "test_state": test_state,
+                "test_front": "bad" if fail_first else "good",
+            },
+        )
+        .build()
+    )
+    return dr
+
+
+def test_parallel_graceful_simple() -> None:
+    # Fail before the anything is yielded: should get 1 failure and skip all
+    dr = _make_driver(
+        temp_module_simple,
+        "early",
```

Comment:
These `test_state`, `find_parallel`, and the expected value would be a good use for `@parameterize` -- it'll break it into future tests and make finding the bug one introduces easier. LMK if you want help. figuring that out! Thinking something like:


```python
@pytest.mark.parameterize(
    "test_state,find_parallel,fail_first,expected_value", # add more
    ["early", False, False, [None]]
)
def test_parallel_graceful_simple(test_state,find_parallel,fail_first,expected_value):
    dr = _make_driver(temp_module_simple, test_state, find_prallel=find_parallel, fail_first=fail_first)
    assert dr.execute(["distro_gather"], inputs={"n" : 10}) == expected_value
```

(but modify a bit, just jotting down some ideas)